test_launch_ros/test/test_launch_ros/test_normalize_parameters.py

```diff
@@ -0,0 +1,173 @@
+
+def test_dictionary_with_substitution_list_name():
+    orig = [{(TextSubstitution(text='bar'), TextSubstitution(text='foo')): 1}]
+    norm = normalize_parameters(orig)
+    expected = ({'barfoo': 1},)
+    assert evaluate_parameters(LaunchContext(), norm) == expected
+
+
+def test_dictionary_with_substitution_list_value():
+    orig = [{'foo': [TextSubstitution(text='fiz'), TextSubstitution(text='buz')]}]
+    norm = normalize_parameters(orig)
+    expected = ({'foo': 'fizbuz'},)
+    assert evaluate_parameters(LaunchContext(), norm) == expected
+
+    orig = [{'foo': [[TextSubstitution(text='fiz')], [TextSubstitution(text='buz')]]}]
+    norm = normalize_parameters(orig)
+    expected = ({'foo': ('fiz', 'buz')},)
+    assert evaluate_parameters(LaunchContext(), norm) == expected
+
+
+def test_dictionary_with_mixed_substitutions_and_strings():
+    orig = [{'foo': [TextSubstitution(text='fiz'), 'bar']}]
+    norm = normalize_parameters(orig)
+    expected = ({'foo': 'fizbar'},)
+    assert evaluate_parameters(LaunchContext(), norm) == expected
+
+    orig = [{'foo': [[TextSubstitution(text='fiz')], 'bar']}]
+    norm = normalize_parameters(orig)
+    expected = ({'foo': ('fiz', 'bar')},)
+    assert evaluate_parameters(LaunchContext(), norm) == expected
+
+
+def test_dictionary_with_str():
+    orig = [{'foo': 'bar', 'fiz': ['b', 'u', 'z']}]
+    norm = normalize_parameters(orig)
+    expected = ({'foo': 'bar', 'fiz': ('b', 'u', 'z')},)
+    assert evaluate_parameters(LaunchContext(), norm) == expected
+
+
+def test_dictionary_with_bool():
+    orig = [{'foo': False, 'fiz': [True, False, True]}]
+    norm = normalize_parameters(orig)
+    expected = ({'foo': False, 'fiz': (True, False, True)},)
+    assert evaluate_parameters(LaunchContext(), norm) == expected
+
+
+def test_dictionary_with_float():
+    orig = [{'foo': 1.2, 'fiz': [2.3, 3.4, 4.5]}]
+    norm = normalize_parameters(orig)
+    expected = ({'foo': 1.2, 'fiz': (2.3, 3.4, 4.5)},)
+    assert evaluate_parameters(LaunchContext(), norm) == expected
+
+
+def test_dictionary_with_int():
+    orig = [{'foo': 1, 'fiz': [2, 3, 4]}]
+    norm = normalize_parameters(orig)
+    expected = ({'foo': 1, 'fiz': (2, 3, 4)},)
+    assert evaluate_parameters(LaunchContext(), norm) == expected
+
+
+def test_dictionary_with_int_and_float():
+    orig = [{'foo': 1, 'fiz': [2, 3.1, 4]}]
+    norm = normalize_parameters(orig)
+    expected = ({'foo': 1, 'fiz': (2.0, 3.1, 4.0)},)
+    evaluated = evaluate_parameters(LaunchContext(), norm)
+    assert evaluated == expected
+    # pytest doesn't check int vs float type
+    assert tuple(map(type, evaluated[0]['fiz'])) == (float, float, float)
+
+
+def test_dictionary_with_bytes():
+    orig = [{'foo': 1, 'fiz': bytes([0xff, 0x5c, 0xaa])}]
+    norm = normalize_parameters(orig)
+    expected = ({'foo': 1, 'fiz': bytes([0xff, 0x5c, 0xaa])},)
+    assert evaluate_parameters(LaunchContext(), norm) == expected
+
+
+def test_dictionary_with_dissimilar_array():
```

Comment:
I'm curious to know if a dissimilar array with a substitution type will also work. In particular:

```
[True, 1, TextSubstitution(text='foo')]
```
and
```
[TextSubstitution(text='foo'), True, 1]
```

Reply:
<details><summary>Good catch, there's a bug</summary>

```
test/test_launch_ros/test_normalize_parameters.py:180: in test_dictionary_with_dissimilar_array
    norm = normalize_parameters(orig)
../../../../build/launch_ros/launch_ros/utilities/normalize_parameters.py:183: in normalize_parameters
    normalized_params.append(normalize_parameter_dict(param))
../../../../build/launch_ros/launch_ros/utilities/normalize_parameters.py:162: in normalize_parameter_dict
    normalized[tuple(name)] = _normalize_parameter_array_value(value)
../../../../build/launch_ros/launch_ros/utilities/normalize_parameters.py:84: in _normalize_parameter_array_value
    return tuple(normalize_to_list_of_substitutions(cast(SomeSubstitutionsType, value)))
../../../../build/launch/launch/utilities/normalize_to_list_of_substitutions_impl.py:42: in normalize_to_list_of_substitutions
    return [normalize(y) for y in cast(Iterable, subs)]
../../../../build/launch/launch/utilities/normalize_to_list_of_substitutions_impl.py:42: in <listcomp>
    return [normalize(y) for y in cast(Iterable, subs)]
../../../../build/launch/launch/utilities/normalize_to_list_of_substitutions_impl.py:36: in normalize
    "'str' or 'launch.Substitution' were expected.".format(type(x)))
E   TypeError: Failed to normalize given item of type '<class 'bool'>', when only 'str' or 'launch.Substitution' were expected.
```
<details>

Reply:
Added tests and fixed bug in a4ac02b6c423f3dccd3d90be97d805b9e97239db